src/encrypter.c

```diff
@@ -17,74 +17,135 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include "getopt.h"
 #include "config.h"
 #include "crypto.h"
+#include "checksum.h"
 
 
+static void exithelp(void) {
+    printf("Usage: fwcrypt [options] infile outfile\n"
+           "\t -e Encrypt (default)\n"
+           "\t -d Decrypt\n"
+           "\t -c With checksum signature\n"
+    );
+    exit(0);
+}
 
-static const char *shelp = "Usage: encrypter [option] infile outfile\n" \
-                           "\t -e Encrypt file\n" \
-                           "\t -d Decrypt file\n";
 
-static const char *sferr = "Error: unable to open file %s\n";
+int main(int argc, char **argv)
+{
+    int dir = 1;
+    int crc = 0;
+    int c;
+
+    opterr = 0;
+
+    while ((c = getopt(argc, argv, "edch")) != -1)
+        switch (c)
+        {
+        case 'e':
+            dir = 1;
+            break;
+        case 'd':
+            dir = 0;
+            break;
+        case 'c':
+            crc = 1;
+            break;
+        case 'h':
+        case '?':
+            exithelp();
+            break;
+        default:
+            exit(-1);
+        }
 
-static int help(void) {
-    printf(shelp);
-    return -1;
-}
+    if ((argc - optind) != 2) exithelp();
 
-static int file_crypt(char *stro, char *stri, int dir) {
-    uint32_t buff[0x100];
-    uint32_t bytes = 0;
-    FILE *fi = fopen(stri, "rb");
+    FILE *fi = fopen(argv[optind], "rb");
     if (fi == NULL) {
-        printf(sferr, stri);
-        return 1;
+        printf("Failed to open file: %s\n", argv[optind]);
+        exit(1);
     }
-    FILE *fo = fopen(stro, "wb");
+
+    optind++;
+
+    FILE *fo = fopen(argv[optind], "wb");
     if (fo == NULL) {
         fclose(fi);
-        printf(sferr, stro);
-        return 2;
+        printf("Failed to open file: %s\n", argv[optind]);
+        exit(2);
     }
-    aes_init();
-    do {
-        size_t inbytes = fread(buff, 1, sizeof(buff), fi);
 
-    inbytes = ((inbytes + (CRYPTO_BLKSIZE - 1)) / CRYPTO_BLKSIZE) * CRYPTO_BLKSIZE;
+    fseek(fi, 0, SEEK_END);
+    uint32_t length = ftell(fi);
+    fseek(fi, 0, SEEK_SET);
 
-#if defined(DFU_USE_CIPHER)
-
-        if (dir == 1) {
-            aes_encrypt(buff, buff, (int)inbytes);
-        } else {
-            aes_decrypt(buff, buff, (int)inbytes);
-        }
-#endif
-        bytes += inbytes;
-        size_t outbytes = fwrite(buff, 1, inbytes, fo);
-
-    } while (!feof(fi));
+    uint32_t *buf = malloc(length + 20);
+    if (buf == NULL) {
+        printf("Failed to allocate buffer. length %d\n", length);
+    }
 
+    fread(buf, 1, length, fi);
     fclose(fi);
-    fclose(fo);
 
+    aes_init();
+    if (dir) {
+#if (DFU_VERIFY_CHECKSUM != _DISABLE)
+        if (crc) {
+            uint32_t cs = calculate_checksum(buf, length);
+            printf("Firmware lendth: %d, checksum: %08X\n"
+                   "Addind signatute.\n",
```

Comment:
Typo: length, Adding and signature

Reply:
Fixed https://github.com/dmitrystu/sboot_stm32/pull/10/commits/682b712afdb106398984adf115e3269a534b2684